Feat: parse translated challenges

ojeytonwilliams opened the pull request:
This treats translated challenges differently from English ones.  It takes the instructions, descriptions and test `text` from the translated challenge.  However, the seed, solution and `testString`s come from the English.  This means that bug fixes are automatically applied to all translations and translators can concentrate on language, without worrying about code.

Comments in the seed are replaced with translated versions if they exist.

Video challenges get the question and answers from the translated challenge.

Need to do
- [x] Create a real comment translation 'dictionary' @RandellDawson is working on this (done for Responsive Web Design)
- [x] ~~Make builds fail if any problems occur during parsing (missing comments only result in a warning, everything else should terminate the build)~~ the build no longer fails, so that the auditing can continue, but this should be reinstated before going to production
- [x] Remove the FOR TESTING commit
- [x] ~~`npm test:curriculum` should ignore locale, since all challenges will use the English tests (actually it had its own environment variable and the tests can be run using the results of `parseTranslation`)~~ I went in the opposite direction - tests are now controlled by `LOCALE` and can be run independently.  They should not fail if the English tests pass, but this should warn us if there are mismatches (challenges that do not appear in all languages, for example)
- [x] certificates need handling separately
- [ ] ~~check if `.language.md` extensions are necessary, if not https://github.com/freeCodeCamp/freeCodeCamp/pull/39024 can be closed.~~ To avoid changing too much in one go, I'm sticking with them for now.

Would be nice
- [x] Warn if a comment is not in the dictionary

Closes #37955
Closes #37599

ojeytonwilliams commented:
This is mostly a write-up of what @ahmadabdolsaheb and I discussed.  

Having translated challenges with descriptions that don't match the tests is something we'd need to avoid.  This could happen if the English challenge gets updated with a new description and tests (and no-one updates the translations).  As soon as the changes to the English are live, the translated challenges would start using the new tests and there would be a mismatch.

One approach would be to automatically create translation issues whenever an English challenge's text changes (via Github Actions, maybe) and only allow the changes into production once all the translations are in.

A more elaborate (and difficult to implement) approach would be to version the challenges and tie the translations to a particular version. That way a new English version could be quickly pushed to live, with the translations using the tests from the old version until a new translation PR is merged.

ojeytonwilliams commented:
To expand a bit, I think this PR would cause most problems when the `text` *and* `testString` are both updated in the English.  Reason being, this is likely a change in the requirements to complete the challenge, not simply a bug fix.

Any English language PR which changes neither the `testString`s nor the seed has no affect the translated challenges at all, so is not breaking.  An issue should be created, since it's a hint that the translations may need changing, but the English PR can be merged regardless.

If the description and the seed are changed, it could mean that the translated description now refers to something that has been removed from the seed.  This would be breaking.

In summary:

Seed change or `testString` change combined with any other text changes could be breaking. Everything else should not be.

ojeytonwilliams committed:
test: add tests for translated challenge parser

ojeytonwilliams committed:
feat: parse translated challenges

Using the English challenge as a source for the seed, solution and tests
this takes the parts that can be translated from the translated version
of the challenge. It also translates known comments in the seed.

ojeytonwilliams commented:
@raisedadead I've cleaned it up, given it another look and it seems fine.  It probably should target the `next-chinese` branch, right?  If so, is there any reason to keep a PR targeting master?

raisedadead commented:
> It probably should target the `next-chinese` branch, right? If so, is there any reason to keep a PR targeting master?

Well if you are able to rebase this on top of `next-chinese`, then sure, if not you can leave it as is.

ojeytonwilliams committed:
fix: remove cn challenges already removed from eng

ojeytonwilliams commented:
@raisedadead I had to remove 10 Chinese challenges that were no longer in the English, but having done that, the Chinese curriculum only has 1467 challenges (to the 1533 English ones).

Also, there are a few challenges where the tests don't match.  For now I've suppressed the warnings so the site can be built.

RandellDawson commented:
> the Chinese curriculum only has 1467 challenges (to the 1533 English ones).

Do you have a list of the English challenges that need a machine translation, so the Chinese team can review and make suggestions for correction?  I can do the translations for those challenges in a separate PR or we can just create an issue listing all the challenges that need to be converted to Chinese and have the Chinese team or other contributors to "real" translations of them.

ojeytonwilliams commented:
> Do you have a list of the English challenges that need a machine translation

Not yet.  I'll make a list of all the challenges with mismatched tests and all the ones that don't exist in the Chinese.

raisedadead commented:
Okay, I understand that the Chinese curriculum has obviously divulged. The idea is to be able to make the translations live one cert at a time. We know that all the Web Design Cert stuff should be in order. And currently the @freeCodeCamp/chinese team is working on the Basic JavaScript sections.

So, we could modify the parser to be able to pick only certs that we choose?

raisedadead commented:
After landing this PR, we would like to be able to land one certificate at a time on the .dev so, that it unblocks the Chinese mods team to start working on the next one in tandem.

ojeytonwilliams commented:
> So, we could modify the parser to be able to pick only certs that we choose?

Absolutely.  So, should it just default to the English unless it's one of the allowed certs?

ojeytonwilliams committed:
feat: restrict translation to audited certs

ojeytonwilliams commented:
The Chinese challenges that are not in the English curriculum, with explanations (most have been renamed):

```
02-javascript-algorithms-and-data-structures/es6/understand-the-differences-between-import-and-require.chinese.md
renamed to
reuse-javascript-code-using-import

02-javascript-algorithms-and-data-structures/es6/use-destructuring-assignment-with-the-rest-operator-to-reassign-array-elements.chinese.md
renamed to
use-destructuring-assignment-with-the-rest-parameter-to-reassign-array-elements

02-javascript-algorithms-and-data-structures/es6/use-export-to-reuse-a-code-block.chinese.md
renamed to 
use-export-to-share-a-code-block

02-javascript-algorithms-and-data-structures/es6/use-the-rest-operator-with-function-parameters.chinese.md
renamed to
use-the-rest-parameter-with-function-parameters

02-javascript-algorithms-and-data-structures/es6/write-concise-object-literal-declarations-using-simple-fields.chinese.md
renamed to
write-concise-object-literal-declarations-using-object-property-shorthand

02-javascript-algorithms-and-data-structures/es6/write-higher-order-arrow-functions.chinese.md
moved to
02-javascript-algorithms-and-data-structures/functional-programming/use-higher-order-functions-map-filter-or-reduce-to-solve-a-complex-problem

03-front-end-libraries/react/manage-updates-with-lifecycle-methods.chinese.md
removed from the curriculum

08-coding-interview-prep/data-structures/remove-from-a-set.chinese.md
removed, merged into
create-a-set-class
```

All the challenges where the number of tests in the English doesn't match the number in the Chinese:
> Apply a Style Until a Condition is Met with @<!-- -->while
Create a Media Query
Add Comments in JSX
Request Header Parser Microservice
Make an Image Responsive
Find Characters with Lazy Matching
Create Strings using Template Literals
Add Placeholder Text to a Text Field
Manage State Locally First
Create a Bulleted Unordered List
Inherit Behaviors from a Supertype
Check if an Object has a Property
Understanding the Differences between the freeCodeCamp and Browser Console
Mutations
Repeat a String Repeat a String
Create a Set Class
Match Characters that Occur Zero or More Times
Understand Own Properties
Use Destructuring Assignment to Assign Variables from Nested Objects
Iterate Through the Keys of an Object with a for...in Statement
Use Destructuring Assignment to Assign Variables from Objects
Use Destructuring Assignment to Pass an Object as a Function's Parameters
Give Sibling Elements a Unique Key Attribute
Steamroller
Use getters and setters to Control Access to an Object
Remove Items Using splice()
Use Closure to Protect Properties Within an Object from Being Modified Externally
Use the reduce Method to Analyze Data
Pass Props to a Stateless Functional Component
Positive and Negative Lookahead
Restrict Possible Usernames
Use Capture Groups to Search and Replace
Remove Elements from a Linked List by Index
Remove Elements from a Linked List
Use Clockwise Notation to Specify the Margin of an Element
Use Array.map() to Dynamically Render Elements
Use CSS Selectors to Style Elements
Escape Sequences in Strings
Use State to Toggle an Element
Find the Length of a String
Use the u Tag to Underline Text
Record Collection
Word Blanks
Problem 2: Even Fibonacci Numbers

Finally, all the challenges that are missing from the Chinese curriculum, this includes some which have been renamed:

>  '02-javascript-algorithms-and-data-structures/regular-expressions/check-for-mixed-grouping-of-characters',
  '02-javascript-algorithms-and-data-structures/es6/complete-a-promise-with-resolve-and-reject',
  '08-coding-interview-prep/data-structures/check-if-binary-search-tree',
  '02-javascript-algorithms-and-data-structures/es6/create-a-javascript-promise',
  '04-data-visualization/json-apis-and-ajax/get-json-with-the-javascript-fetch-method',
  '02-javascript-algorithms-and-data-structures/es6/create-a-module-script',
  '02-javascript-algorithms-and-data-structures/es6/handle-a-fulfilled-promise-with-then',
  '02-javascript-algorithms-and-data-structures/es6/handle-a-rejected-promise-with-catch',
  '02-javascript-algorithms-and-data-structures/es6/reuse-javascript-code-using-import',
  '02-javascript-algorithms-and-data-structures/functional-programming/use-higher-order-functions-map-filter-or-reduce-to-solve-a-complex-problem',
  '08-coding-interview-prep/rosetta-code/cumulative-standard-deviation',
  '08-coding-interview-prep/rosetta-code/cusip',
  '02-javascript-algorithms-and-data-structures/es6/use-destructuring-assignment-to-extract-values-from-objects',
  '08-coding-interview-prep/rosetta-code/cut-a-rectangle',
  '02-javascript-algorithms-and-data-structures/es6/use-destructuring-assignment-with-the-rest-parameter-to-reassign-array-elements',
  '02-javascript-algorithms-and-data-structures/es6/use-export-to-share-a-code-block',
  '02-javascript-algorithms-and-data-structures/es6/use-the-rest-parameter-with-function-parameters',
  '02-javascript-algorithms-and-data-structures/es6/write-concise-object-literal-declarations-using-object-property-shorthand',
  '08-coding-interview-prep/rosetta-code/dot-product',
  '08-coding-interview-prep/rosetta-code/generator-exponential',
  '08-coding-interview-prep/rosetta-code/k-d-tree',
  '08-coding-interview-prep/rosetta-code/kaprekar-numbers',
  '02-javascript-algorithms-and-data-structures/basic-javascript/replace-loops-using-recursion',
  '08-coding-interview-prep/rosetta-code/knights-tour',
  '08-coding-interview-prep/rosetta-code/largest-int-from-concatenated-ints',
  '08-coding-interview-prep/rosetta-code/last-friday-of-each-month',
  '08-coding-interview-prep/rosetta-code/leap-year',
  '08-coding-interview-prep/rosetta-code/least-common-multiple',
  '08-coding-interview-prep/rosetta-code/left-factorials',
  '08-coding-interview-prep/rosetta-code/sort-an-array-of-composite-structures',
  '08-coding-interview-prep/rosetta-code/sort-disjoint-sublist',
  '08-coding-interview-prep/rosetta-code/sort-stability',
  '08-coding-interview-prep/rosetta-code/sort-using-a-custom-comparator',
  '08-coding-interview-prep/rosetta-code/sorting-algorithms-bead-sort',
  '08-coding-interview-prep/rosetta-code/sorting-algorithms-bogosort',
  '08-coding-interview-prep/rosetta-code/sorting-algorithms-cocktail-sort',
  '08-coding-interview-prep/rosetta-code/sorting-algorithms-comb-sort',
  '08-coding-interview-prep/rosetta-code/sorting-algorithms-gnome-sort',
  '08-coding-interview-prep/rosetta-code/sorting-algorithms-pancake-sort',
  '08-coding-interview-prep/rosetta-code/sorting-algorithms-permutation-sort',
  '08-coding-interview-prep/rosetta-code/sorting-algorithms-shell-sort',
  '08-coding-interview-prep/rosetta-code/sorting-algorithms-stooge-sort',
  '08-coding-interview-prep/rosetta-code/sorting-algorithms-strand-sort',
  '08-coding-interview-prep/rosetta-code/soundex',
  '08-coding-interview-prep/rosetta-code/spiral-matrix',
  '02-javascript-algorithms-and-data-structures/basic-javascript/use-recursion-to-create-a-range-of-numbers',
  '08-coding-interview-prep/rosetta-code/split-a-character-string-based-on-change-of-character',
  '08-coding-interview-prep/rosetta-code/state-name-puzzle',
  '08-coding-interview-prep/rosetta-code/stern-brocot-sequence',
  '08-coding-interview-prep/rosetta-code/straddling-checkerboard',
  '08-coding-interview-prep/rosetta-code/stream-merge',
  '08-coding-interview-prep/rosetta-code/strip-control-codes-and-extended-characters-from-a-string',
  '08-coding-interview-prep/rosetta-code/subleq',
  '08-coding-interview-prep/rosetta-code/sudoku',
  '08-coding-interview-prep/rosetta-code/sum-digits-of-an-integer',
  '08-coding-interview-prep/rosetta-code/sum-multiples-of-3-and-5',
  '08-coding-interview-prep/rosetta-code/sum-of-a-series',
  '08-coding-interview-prep/rosetta-code/sum-of-squares',
  '08-coding-interview-prep/rosetta-code/sum-to-100',
  '08-coding-interview-prep/rosetta-code/sutherland-hodgman-polygon-clipping',
  '08-coding-interview-prep/rosetta-code/symmetric-difference'

Apologies for the slight differences in naming (titles, paths, etc), I can be more consistent if needed, but they should all be searchable!

ojeytonwilliams commented:
I've noticed some challenge files are not in the form `/path/to/name.lang.md` e.g. 
```
curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/stand-in-line.english.md
```
is fine, but
```
curriculum/challenges/english/08-coding-interview-prep/rosetta-code/knapsack-problem-0-1.md
```
is not.

I'm updating my PR to throw errors when it encounters these and I'll create PRs fixing them, but some of the above lists may have false positives.  I'll update the lists ASAP.

RandellDawson commented:
All files should have the language extension + `md`.

RandellDawson commented:
@ankush0512  Just wanted to let you know, that I have already have PRs that consolidate the comments (they are currently being reviewed) and will be merged shortly.  I have also created the applicable translation lookup object based on the consolidated comments that were left.  I will also be taking care of the other curriculum sections as we will be  working through one section at a time.

RandellDawson committed:
fix: added Chinese comment translations for Responsive Web Design

raisedadead commented:
I bet we are getting along real close with this? 

I have landed all the Chinese i18n progress so far (for the first cert). I have also rebased this PR off of master and changed the target.

ojeytonwilliams commented:
> I bet we are getting along real close with this?

Yep, pretty close.  I need to review Randell's PRs though, so I'll get on that.

ojeytonwilliams commented:
@raisedadead btw, do you want the un-validated certs (i.e. non Responsive Web Design) to disappear from the curriculum entirely or just default to English?

raisedadead commented:
I would be fine with them disappearing.

ojeytonwilliams commented:
> I would be fine with them disappearing.

I'm looking into this, but it has to be done carefully or Gatsby breaks.  Might take a little while.

ojeytonwilliams committed:
fix: add explicit schema for hidden challenges

raisedadead commented:
> I'm looking into this, but it has to be done carefully or Gatsby breaks. Might take a little while.

OK, I would defer to your best judgement. If it becomes too annoying to implement, let us default to English.

Let me know if I can be of any help.

ojeytonwilliams commented:
@raisedadead> Let me know if I can be of any help.

I think it's more or less okay. The last two commits do *most* of it, but there's still the issue that all the certs appear in the settings page, even though they're gone from the curriculum.  Finally I still need to handle the certificate challenges, as they have a different structure for the tests, but it should be easy to do.

If you could take a look and let me know what needs doing differently, that would be a huge help.

RandellDawson commented on client/plugins/fcc-source-challenges/create-challenge-nodes.js:
```diff
@@ -29,7 +30,43 @@ function createChallengeNode(challenge, reporter) {
+            ? [{ link: '', src: '' }]
+            : challenge.required,
+          files: isEmpty(challenge.files)
+            ? {
```
What about creating an object (i.e. `emptyFiles`) and stick it after the `?`.  Then before the return statement, you could use something like the following to create the object.
```
const emptyFiles = ['html', 'js', 'jsx'].reduce((obj, type) => {
  const properties = ['key', 'ext', 'name', 'contents', 'head', 'tail'];
  const keyValuesObj = properties.reduce((obj, prop) => ({ ...obj, [prop]: ''}), {});
  return { ...obj, [`index${type}`]: keyValuesObj };
}, {});
```

ojeytonwilliams replied:
Good point, it is rather ugly.  I'll sort that out tomorrow!

ojeytonwilliams replied:
I based my solution on yours, but decided against the nested reduce for reasons of readability.  Is this version okay?

ojeytonwilliams committed:
test: add certificate specific tests

ojeytonwilliams committed:
fix: handle certificate challenges separately

RandellDawson commented on client/plugins/fcc-source-challenges/create-challenge-nodes.js:
```diff
@@ -1,6 +1,20 @@
+  acc[curr] = { ...FILE_CONTENTS };
+  return acc;
+}, {});
+
```
```suggestion
const EMPTY_FILES = ['indexhtml', 'indexjs', 'indexjsx'].reduce(
  (acc, curr) => ({ ...acc, [curr]: FILE_CONTENTS })
, {});
```

ojeytonwilliams replied:
I like it.  Could you put this comment on the latest commit?  Github's telling me I can't commit the suggestion because it's 'outdated'.

ojeytonwilliams replied:
When I cleaned things up, I included this with one last tweak: it now clones `FILE_CONTENTS`.

RandellDawson reviewed (changes requested):
How about a slightly more concise version?

ojeytonwilliams commented:
@raisedadead the basic functionality is all there, for the first cert at least.  I went ahead and removed the untranslated/unaudited certs from the settings, in the last commit.

RandellDawson commented:
@ojeytonwilliams So is the plan to merge this one and make sure nothing "breaks" on the dev site and then remove all but the Description, Instructions, and `text`s of the Tests section in a separate PR and make sure nothing breaks?

ojeytonwilliams commented:
The way I see it going is this:

1. I'll clean it up (there are way too many commits at the moment)
2. Test it locally
3. Test on .dev (or a Chinese specific domain?)

Removing the unnecessary parts from the Chinese challenges can happen whenever we want. The parser can already cope, the latest test just confirms that. Presumably we'll remove them after we've used it in production for a while and haven't run into any insurmountable problems.

raisedadead commented:
> Removing the unnecessary parts from the Chinese challenges can happen whenever we want.

We have a snapshot already (`backup-*` branches). However, I would still want to do it in a separate PR.

raisedadead commented:
Also, I am ready to pull this down and deploy whenever you want me to. I will prepare the Vm's etc, when you give me the go ahead.

ojeytonwilliams commented:
@raisedadead it should be good to go. There was a bug, now squashed, where the English settings only showed Responsive Web Design, so it should be QAed in both English and Chinese.  Just to be sure.

raisedadead commented:
I have started the work on deploying this at `chinese.freecodecamp.dev` via https://github.com/freeCodeCamp/nginx-config/pull/1, I will have something by morning.

ojeytonwilliams commented:
I avoided using a parser in this because I was worried that it would reformat the code, but 
 @RandellDawson's been looking at using Babel and a css parser to remove comments from user code.  If that works and has no unwanted side-effects, we can consider using that instead of the regex.

moT01 commented:
I have blocked this for now. See the [pinned code freeze issue](https://github.com/freeCodeCamp/freeCodeCamp/issues/38308)

ojeytonwilliams committed:
fix: update chinese certificates

ojeytonwilliams committed:
fix: robust Certifications/Additional learning split

ojeytonwilliams committed:
fix(curriculum): add missing .english extensions

ojeytonwilliams committed:
fix: QA/Infosec update and python to chinese

ojeytonwilliams committed:
fix: add missing chinese extensions

ojeytonwilliams committed:
fix: test text mismatched quotes

ojeytonwilliams committed:
fix: handle video challenges

ojeytonwilliams committed:
fix: simplify language specific testing

Prevents multiple languages' challenges from being tested at once.

ojeytonwilliams committed:
fix: relax schema while auditing translations

Without this change we would have to put in dummy localeTitle values.
This defeats the point of making it required.

ojeytonwilliams committed:
fix: insert blank line after ```

search and replace ```\n< with ```\n\n< to ensure there's an empty line
before closing tags

ojeytonwilliams committed:
fix: outstanding md linting issues
